File: app/src/main/java/engine/core/RenderThread.java
package engine.core;

import java.util.concurrent.atomic.AtomicReference;

import engine.graphics.DrawCall;

public class RenderThread extends Thread {

    private final EngineContext engine;
    private DisplayManager displayManager;
    private boolean running;

    public RenderThread(EngineContext engine) {
        super();
        this.engine = engine;
    }

    @Override
    public synchronized void start() {
        running = true;
        displayManager = this.engine.getDisplayManager();
        super.start();
    }

    private final AtomicReference<DrawCall> readyFrame = new AtomicReference<>(null);
    private final Object notifyHandler = new Object();

    public void submitFrame(DrawCall frame) {
        DrawCall previousFrame = readyFrame.getAndSet(frame);

        synchronized (notifyHandler) {
            notifyHandler.notify();
        }

        if (previousFrame != null) {
            engine.getPerformanceMonitor().droppedFrames.incrementAndGet();
        }
    }

    @Override
    public void run() {
        try {
            long lastTime = System.nanoTime();
            long idleTime = 0;
            while (!isInterrupted()) {
                DrawCall frame = readyFrame.getAndSet(null);
                if (frame == null) {
                    idleTime = System.nanoTime();
                    synchronized (notifyHandler) {
                        notifyHandler.wait();
                    }
                    idleTime = System.nanoTime() - idleTime;
                    continue;
                }


                long renderBeginTime = System.nanoTime();
                displayManager.render(frame);
                long renderEndTime = System.nanoTime();

                engine.getPerformanceMonitor().drawTime.set(PerformanceMonitor.nanosToMicros(renderEndTime - renderBeginTime));
                engine.getPerformanceMonitor().drawDelta.set(PerformanceMonitor.nanosToMicros(renderBeginTime - lastTime));
                engine.getPerformanceMonitor().drawIdle.set(PerformanceMonitor.nanosToMicros(idleTime));

                idleTime = 0;
                lastTime = renderBeginTime;
            }
        } catch (InterruptedException ignored) {}
    }
}

File: engine2d/src/main/java/life/unmedicated/engine2d/core/RenderThread.kt
package life.unmedicated.engine2d.core

import kotlinx.coroutines.internal.SynchronizedObject
import life.unmedicated.engine2d.graphics.DrawCall
import java.util.concurrent.CountDownLatch
import java.util.concurrent.SynchronousQueue
import java.util.concurrent.atomic.AtomicReference
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class RenderThread(private val engineContext: EngineContext) : Thread() {
    private val latchRef = AtomicReference<CountDownLatch>()
    private val readyFrame = AtomicReference<DrawCall>()

    fun submitFrame(frame: DrawCall) {
        readyFrame.getAndSet(frame)
        synchronized(latchRef) {
            latchRef.getAndSet(null)?.countDown()
        }
    }

    override fun run() {
        try {
            while (!isInterrupted) {
                val frame = readyFrame.getAndSet(null)
                if (frame == null) {
                    var latch: CountDownLatch
                    synchronized(latchRef) {
                        latch = CountDownLatch(1)
                        latchRef.set(latch)
                    }
                    latch.await()
                    continue
                }

                engineContext.displayManager.renderFrame(frame)
            }
        } catch (_: InterruptedException) {}
    }
}